crates/task-impls/src/transactions.rs

```diff
@@ -120,9 +120,245 @@ pub struct TransactionTaskState<TYPES: NodeType, I: NodeImplementation<TYPES>> {
+                block_view
+            );
+
+            // Increment the metric for number of empty blocks proposed
+            self.consensus
+                .write()
+                .await
+                .metrics
+                .number_of_empty_blocks_proposed
+                .add(1);
+
+            let membership_total_nodes = self.membership.total_nodes();
+            let Some(null_fee) = null_block::builder_fee(self.membership.total_nodes(), version)
+            else {
+                error!("Failed to get null fee");
+                return None;
+            };
+
+            // Create an empty block payload and metadata
+            let (_, metadata) = <TYPES as NodeType>::BlockPayload::empty();
+
+            let (_, precompute_data) = precompute_vid_commitment(&[], membership_total_nodes);
+
+            // Broadcast the empty block
+            broadcast_event(
+                Arc::new(HotShotEvent::BlockRecv(PackedBundle::new(
+                    vec![].into(),
+                    metadata,
+                    block_view,
+                    vec1::vec1![null_fee.clone()],
+                    vec1::vec1![null_fee],
+                    Some(precompute_data),
+                ))),
+                &event_stream,
+            )
+            .await;
+        };
+
+        return None;
+    }
+
+    /// marketplace view change handler
+    pub async fn handle_view_change_marketplace(
+        &mut self,
+        event_stream: Sender<Arc<HotShotEvent<TYPES>>>,
+        block_view: TYPES::Time,
+    ) -> Option<HotShotTaskCompleted> {
+        let version = match hotshot_types::simple_certificate::version(
+            block_view,
+            &self
+                .decided_upgrade_certificate
+                .read()
+                .await
+                .as_ref()
+                .cloned(),
+        ) {
+            Ok(v) => v,
+            Err(err) => {
+                error!("Upgrade certificate requires unsupported version, refusing to request blocks: {}", err);
+                return None;
+            }
+        };
+
+        // Only request bundles and propose with a nonempty block if we are not between versions.
+        if !self
+            .decided_upgrade_certificate
+            .read()
+            .await
+            .as_ref()
+            .is_some_and(|cert| cert.upgrading_in(block_view))
+        {
+            if let Ok(bundles) = self
+                .auction_results_provider
+                .fetch_bundles(block_view)
+                .await
+            {
+                let Some(null_fee) =
+                    null_block::builder_fee(self.membership.total_nodes(), version)
+                else {
+                    error!("Failed to get null fee");
+                    return None;
+                };
+
+                let mut sequencing_fees = Vec::new();
+                let mut bid_fees: Vec<BuilderFee<TYPES>> = Vec::new();
```

Comment:
We'll want to remove the `bid_fee` variable eventually.

Reply:
Removed!